necessary `unsafe` keywords marked as unnecessary when inside incomplete match

### Code

```Rust
enum MyEnum {
    Red,
    Blue,
    Purple,
    Gray,
    Orange,
}

fn my_fn(input: MyEnum) -> u32 {
    let bytes1 = [1u8, 2u8, 3u8, 4u8];
    let bytes2 = [1u8, 3u8, 3u8, 4u8];
    let bytes3 = [1u8, 2u8, 0u8, 4u8];
    let bytes4 = [1u8, 7u8, 3u8, 4u8];
    match input {
        MyEnum::Red => unsafe { core::mem::transmute_copy(&bytes1) },
        MyEnum::Blue => unsafe { core::mem::transmute_copy(&bytes2) },
        MyEnum::Purple => unsafe { core::mem::transmute_copy(&bytes3) },
        MyEnum::Gray => unsafe { core::mem::transmute_copy(&bytes4) },
    }
}
```


### Current output

```Shell
error[E0004]: non-exhaustive patterns: `MyEnum::Orange` not covered
  --> src/lib.rs:15:11
   |
15 |     match input {
   |           ^^^^^ pattern `MyEnum::Orange` not covered
   |
note: `MyEnum` defined here
  --> src/lib.rs:7:5
   |
2  | enum MyEnum {
   |      ------
...
7  |     Orange,
   |     ^^^^^^ not covered
   = note: the matched value is of type `MyEnum`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
   |
19 ~         MyEnum::Gray => unsafe { core::mem::transmute_copy(&bytes4) },
20 ~         MyEnum::Orange => todo!(),
   |

warning: unnecessary `unsafe` block
  --> src/lib.rs:16:24
   |
16 |         MyEnum::Red => unsafe { core::mem::transmute_copy(&bytes1) },
   |                        ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default

warning: unnecessary `unsafe` block
  --> src/lib.rs:17:25
   |
17 |         MyEnum::Blue => unsafe { core::mem::transmute_copy(&bytes2) },
   |                         ^^^^^^ unnecessary `unsafe` block

warning: unnecessary `unsafe` block
  --> src/lib.rs:18:27
   |
18 |         MyEnum::Purple => unsafe { core::mem::transmute_copy(&bytes3) },
   |                           ^^^^^^ unnecessary `unsafe` block

warning: unnecessary `unsafe` block
  --> src/lib.rs:19:25
   |
19 |         MyEnum::Gray => unsafe { core::mem::transmute_copy(&bytes4) },
   |                         ^^^^^^ unnecessary `unsafe` block

For more information about this error, try `rustc --explain E0004`.
warning: `playground` (lib) generated 4 warnings
error: could not compile `playground` (lib) due to previous error; 4 warnings emitted
```


### Desired output

```Shell
error[E0004]: non-exhaustive patterns: `MyEnum::Orange` not covered
  --> src/lib.rs:15:11
   |
15 |     match input {
   |           ^^^^^ pattern `MyEnum::Orange` not covered
   |
note: `MyEnum` defined here
  --> src/lib.rs:7:5
   |
2  | enum MyEnum {
   |      ------
...
7  |     Orange,
   |     ^^^^^^ not covered
   = note: the matched value is of type `MyEnum`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
   |
19 ~         MyEnum::Gray => unsafe { core::mem::transmute_copy(&bytes4) },
20 ~         MyEnum::Orange => todo!(),
   |
For more information about this error, try `rustc --explain E0004`.
warning: `playground` (lib) generated 4 warnings
error: could not compile `playground` (lib) due to previous error; 4 warnings emitted
```


### Rationale and extra context

Each of the above `unsafe` keywords gets marked with an "unnecessary" warning until the missing match arms are added at which point the warnings disappear. This is confusing since many will see those warnings and remove the `unsafe` blocks only to need to re-add them when they finish writing their match.

Originally reported on RA here: https://github.com/rust-lang/rust-analyzer/issues/15514#issue-1867339857

### Other cases

_No response_

### Anything else?

Any scenario where you have a match statement with necessary `unsafe` blocks in any of the arms, and the match statement is incomplete.

--- a/compiler/rustc_mir_transform/src/check_unsafety.rs
+++ b/compiler/rustc_mir_transform/src/check_unsafety.rs
@@ -483,7 +483,7 @@ fn unsafety_check_result(tcx: TyCtxt<'_>, def: LocalDefId) -> &UnsafetyCheckResu
     // `mir_built` force this.
     let body = &tcx.mir_built(def).borrow();
 
-    if body.is_custom_mir() {
+    if body.is_custom_mir() || body.tainted_by_errors.is_some() {
         return tcx.arena.alloc(UnsafetyCheckResult {
             violations: Vec::new(),
             used_unsafe_blocks: Default::default(),

--- a/tests/ui/unsafe/issue-115348-false-positive-warning-of-unnecessary-unsafe.rs
+++ b/tests/ui/unsafe/issue-115348-false-positive-warning-of-unnecessary-unsafe.rs
@@ -1,0 +1,16 @@
+// Regression test for #115348.
+
+unsafe fn uwu() {}
+
+// Tests that the false-positive warning "unnecessary `unsafe` block"
+// should not be reported, when the error "non-exhaustive patterns"
+// appears.
+
+fn foo(x: Option<u32>) {
+    match x {
+        //~^ ERROR non-exhaustive patterns: `None` not covered
+        Some(_) => unsafe { uwu() },
+    }
+}
+
+fn main() {}

--- a/tests/ui/unsafe/issue-115348-false-positive-warning-of-unnecessary-unsafe.stderr
+++ b/tests/ui/unsafe/issue-115348-false-positive-warning-of-unnecessary-unsafe.stderr
@@ -1,0 +1,21 @@
+error[E0004]: non-exhaustive patterns: `None` not covered
+  --> $DIR/issue-115348-false-positive-warning-of-unnecessary-unsafe.rs:10:11
+   |
+LL |     match x {
+   |           ^ pattern `None` not covered
+   |
+note: `Option<u32>` defined here
+  --> $SRC_DIR/core/src/option.rs:LL:COL
+  ::: $SRC_DIR/core/src/option.rs:LL:COL
+   |
+   = note: not covered
+   = note: the matched value is of type `Option<u32>`
+help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
+   |
+LL ~         Some(_) => unsafe { uwu() },
+LL ~         None => todo!(),
+   |
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0004`.

